File: sdk/python/symbolchain/nem/TransactionFactory.py
from binascii import hexlify

from symbolchain import nc

from ..CryptoTypes import Hash256, PublicKey
from ..RuleBasedTransactionFactory import RuleBasedTransactionFactory
from .Network import Address


class TransactionFactory:
	"""Factory for creating NEM transactions."""

	def __init__(self, network, type_rule_overrides=None):
		"""Creates a factory for the specified network."""
		self.factory = self._build_rules(type_rule_overrides)
		self.network = network

	@staticmethod
	def lookup_transaction_name(transaction_type, transaction_version):
		"""Looks up the friendly name for the specified transaction."""
		return f'{str(transaction_type)[str(transaction_type).index(".") + 1:].lower()}_transaction_v{transaction_version}'

	def create(self, transaction_descriptor, autosort=True):
		"""
		Creates a transaction from a transaction descriptor.
		When autosort is set (default), descriptor arrays requiring ordering will be automatically sorted.
		When unset, descriptor arrays will be presumed to be already sorted.
		"""
		transaction = self.factory.create_from_factory(nc.TransactionFactory.create_by_name, {
			**transaction_descriptor,
			'network': self.network.identifier
		})
		if autosort:
			transaction.sort()

		# hack: explicitly translate transfer message
		if nc.TransactionType.TRANSFER == transaction.type_ and transaction.message and isinstance(transaction.message.message, str):
			transaction.message.message = transaction.message.message.encode('utf8')

		return transaction

	@staticmethod
	def to_non_verifiable_transaction(transaction):
		"""Converts a transaction to a non-verifiable transaction."""
		non_verifiable_class_name = type(transaction).__name__
		if not non_verifiable_class_name.startswith('NonVerifiable'):
			non_verifiable_class_name = f'NonVerifiable{non_verifiable_class_name}'

		non_verifiable_class = getattr(nc, non_verifiable_class_name)
		non_verifiable_transaction = non_verifiable_class()
		for key in dir(non_verifiable_transaction):
			# isupper() to quickly filter out class properties like TRANSACTION_VERSION or TYPE_HINTS
			if key.startswith('_') or key[0].isupper() or key in ('size', 'serialize', 'deserialize') or key.endswith('_computed'):
				continue

			setattr(non_verifiable_transaction, key, getattr(transaction, key))

		return non_verifiable_transaction

	@staticmethod
	def attach_signature(transaction, signature):
		"""Attaches a signature to a transaction."""
		transaction.signature = nc.Signature(signature.bytes)

		transaction_hex = hexlify(TransactionFactory.to_non_verifiable_transaction(transaction).serialize()).decode('utf8').upper()
		signature_hex = str(signature)
		json_payload = f'{{"data":"{transaction_hex}", "signature":"{signature_hex}"}}'
		return json_payload

	@staticmethod
	def _nem_type_converter(value):
		if isinstance(value, Address):
			# yes, unfortunately, nem's Address is 40 bytes string, but we need to pass it as actual bytes not to confuse ByteArray
			return nc.Address(str(value).encode('utf8'))

		return None

	@staticmethod
	def _build_rules(type_rule_overrides):
		factory = RuleBasedTransactionFactory(nc, TransactionFactory._nem_type_converter, type_rule_overrides)
		factory.autodetect()

		struct_names = [
			'CosignatureV1', 'Message', 'NamespaceId', 'MosaicId', 'Mosaic', 'SizePrefixedMosaic', 'MosaicLevy',
			'MosaicProperty', 'SizePrefixedMosaicProperty', 'MosaicDefinition',
			'MultisigAccountModification', 'SizePrefixedMultisigAccountModification', 'SizePrefixedCosignatureV1'
		]
		for name in struct_names:
			factory.add_struct_parser(name)

		sdk_type_mapping = {
			'Address': Address,
			'Hash256': Hash256,
			'PublicKey': PublicKey,
		}
		for name, typename in sdk_type_mapping.items():
			factory.add_pod_parser(name, typename)

		array_names = [
			'SizePrefixedMosaic', 'SizePrefixedMosaicProperty', 'SizePrefixedMultisigAccountModification', 'SizePrefixedCosignatureV1'
		]
		for name in array_names:
			factory.add_array_parser(f'struct:{name}')

		return factory

File: sdk/javascript/src/nem/TransactionFactory.js
import {
	Address,
	/* eslint-disable no-unused-vars */
	Network
	/* eslint-enable no-unused-vars */
} from './Network.js';
import * as nc from './models.js';
import {
	Hash256,
	PublicKey,
	/* eslint-disable no-unused-vars */
	Signature
	/* eslint-enable no-unused-vars */
} from '../CryptoTypes.js';
import RuleBasedTransactionFactory from '../RuleBasedTransactionFactory.js';
import { uint8ToHex } from '../utils/converter.js';

/**
 * Factory for creating NEM transactions.
 */
export default class TransactionFactory {
	/**
	 * Creates a factory for the specified network.
	 * @param {Network} network NEM network.
	 * @param {Map<string, function>|undefined} typeRuleOverrides Type rule overrides.
	 */
	constructor(network, typeRuleOverrides = undefined) {
		/**
		 * @private
		 */
		this._factory = TransactionFactory._buildRules(typeRuleOverrides); // eslint-disable-line no-underscore-dangle

		/**
		 * @private
		 */
		this._network = network;
	}

	/**
	 * Gets rule names with registered hints.
	 */
	get ruleNames() {
		return Array.from(this._factory.rules.keys());
	}

	/**
	 * Looks up the friendly name for the specified transaction.
	 * @param {nc.TransactionType} transactionType Transaction type.
	 * @param {number} transactionVersion Transaction version.
	 * @returns {string} Transaction friendly name.
	 */
	static lookupTransactionName(transactionType, transactionVersion) {
		return `${nc.TransactionType.valueToKey(transactionType.value).toLowerCase()}_transaction_v${transactionVersion}`;
	}

	/**
	 * Creates a transaction from a transaction descriptor.
	 * @param {object} transactionDescriptor Transaction descriptor.
	 * @param {boolean} autosort When set (default), descriptor arrays requiring ordering will be automatically sorted.
	 *                           When unset, descriptor arrays will be presumed to be already sorted.
	 * @returns {nc.Transaction} Newly created transaction.
	 */
	create(transactionDescriptor, autosort = true) {
		const transaction = this._factory.createFromFactory(nc.TransactionFactory.createByName, {
			...transactionDescriptor,
			network: this._network.identifier
		});
		if (autosort)
			transaction.sort();

		// hack: explicitly translate transfer message
		if (nc.TransactionType.TRANSFER === transaction.type && transaction.message && 'string' === typeof (transaction.message.message))
			transaction.message.message = new TextEncoder().encode(transaction.message.message);

		return transaction;
	}

	/**
	 * Converts a transaction to a non-verifiable transaction.
	 * @param {nc.Transaction|nc.NonVerifiableTransaction} transaction Transaction object.
	 * @returns {nc.NonVerifiableTransaction} Non-verifiable transaction object.
	 */
	static toNonVerifiableTransaction(transaction) {
		let nonVerifiableClassName = transaction.constructor.name;
		if (0 !== nonVerifiableClassName.indexOf('NonVerifiable'))
			nonVerifiableClassName = `NonVerifiable${nonVerifiableClassName}`;

		const NonVerifiableClass = nc[nonVerifiableClassName];
		const nonVerifiableTransaction = new NonVerifiableClass();
		Object.getOwnPropertyNames(transaction).forEach(key => {
			if (key in nonVerifiableTransaction)
				nonVerifiableTransaction[key] = transaction[key];
		});

		return nonVerifiableTransaction;
	}

	/**
	 * Attaches a signature to a transaction.
	 * @param {nc.Transaction} transaction Transaction object.
	 * @param {Signature} signature Signature to attach.
	 * @returns {string} JSON transaction payload.
	 */
	static attachSignature(transaction, signature) {
		transaction.signature = new nc.Signature(signature.bytes);

		const transactionHex = uint8ToHex(this.toNonVerifiableTransaction(transaction).serialize());
		const signatureHex = signature.toString();
		const jsonPayload = `{"data":"${transactionHex}", "signature":"${signatureHex}"}`;
		return jsonPayload;
	}

	/**
	 * Tries to coerce an sdk type to a model type.
	 * @param {object} value Value to convert.
	 * @returns {nc.Address|undefined} Converted value or undefined.
	 * @private
	 */
	static _nemTypeConverter(value) {
		if (value instanceof Address) {
			// yes, unfortunately, nem's Address is 40 bytes string, but we need to pass it as actual bytes not to confuse ByteArray
			return new nc.Address(new TextEncoder().encode(value.toString()));
		}

		return undefined;
	}

	/**
	 * Builds a rule based transaction factory.
	 * @param {Map<string, function>|undefined} typeRuleOverrides Type rule overrides.
	 * @returns {RuleBasedTransactionFactory} Rule based transaction factory.
	 * @private
	 */
	static _buildRules(typeRuleOverrides) {
		const factory = new RuleBasedTransactionFactory(nc, this._nemTypeConverter, typeRuleOverrides);
		factory.autodetect();

		[
			'LinkAction', 'MessageType', 'MosaicSupplyChangeAction', 'MosaicTransferFeeType',
			'MultisigAccountModificationType', 'NetworkType', 'TransactionType'
		].forEach(name => { factory.addEnumParser(name); });

		[
			'Message', 'NamespaceId', 'MosaicId', 'Mosaic', 'SizePrefixedMosaic', 'MosaicLevy',
			'MosaicProperty', 'SizePrefixedMosaicProperty', 'MosaicDefinition',
			'MultisigAccountModification', 'SizePrefixedMultisigAccountModification'
		].forEach(name => { factory.addStructParser(name); });

		const sdkTypeMapping = {
			Address,
			Hash256,
			PublicKey
		};
		Object.keys(sdkTypeMapping).forEach(name => { factory.addPodParser(name, sdkTypeMapping[name]); });

		[
			'struct:SizePrefixedMosaic', 'struct:SizePrefixedMosaicProperty', 'struct:SizePrefixedMultisigAccountModification'
		].forEach(name => { factory.addArrayParser(name); });

		return factory;
	}
}
